In [1]:
import pandas as pd
from pathlib import Path

In [2]:
# set path to data file
notebook_dir = Path.cwd()
mover_file = notebook_dir / 'data' / 'SPX_index leavers & joiners_11-Nov-2025.xlsx'
returns_file = notebook_dir / 'data' / 'beginnings.csv'

In [3]:
returns = pd.read_csv(returns_file, parse_dates=['Date'])

In [4]:
returns['Date'] = pd.to_datetime(returns['Date'], format='%Y-%m-%d')
returns = returns[returns['Date'] >= '1995-01-01']
returns.set_index(['Date'], inplace=True)
returns

A.N   AA.N  AABA.OQ^J19  AAL.N^B97  AAL.OQ  AAP.N     AAPL.OQ  \
Date                                                                           
1995-01-03     NaN    NaN          NaN     19.125     NaN    NaN    0.342634   
1995-01-04     NaN    NaN          NaN     19.000     NaN    NaN    0.351562   
1995-01-05     NaN    NaN          NaN     18.625     NaN    NaN    0.347098   
1995-01-06     NaN    NaN          NaN     18.750     NaN    NaN    0.375000   
1995-01-09     NaN    NaN          NaN     18.750     NaN    NaN    0.367885   
...            ...    ...          ...        ...     ...    ...         ...   
2024-12-24  136.10  38.61          NaN        NaN   17.35  43.73  258.200000   
2024-12-26  135.83  38.35          NaN        NaN   17.35  44.24  259.020000   
2024-12-27  135.54  37.68          NaN        NaN   17.35  44.06  255.590000   
2024-12-30  134.42  37.15          NaN        NaN   17.62  46.05  252.200000   
2024-12-31  134.34  37.78          NaN        NaN   17.43  47.29  250.420000   

            ABBV.N  ABI.N^K08  ABK.N^K10  ...  JWN.N^E25        K.N  \
Date                                      ...                         
1995-01-03     NaN    6.56250  12.166667  ...        NaN  27.270216   
1995-01-04     NaN    6.46875  12.333333  ...        NaN  27.328861   
1995-01-05     NaN    6.53125  12.333333  ...        NaN  27.504798   
1995-01-06     NaN    6.46875  12.250000  ...        NaN  27.035633   
1995-01-09     NaN    6.53125  12.333333  ...        NaN  26.859696   
...            ...        ...        ...  ...        ...        ...   
2024-12-24  180.00        NaN        NaN  ...      24.21  80.820000   
2024-12-26  179.20        NaN        NaN  ...      24.22  81.110000   
2024-12-27  178.01        NaN        NaN  ...      24.21  81.170000   
2024-12-30  176.20        NaN        NaN  ...      24.11  80.820000   
2024-12-31  177.70        NaN        NaN  ...      24.15  80.970000   

            KATE.N^G17    KBH.N   KD.N  KDP.OQ    KEY.N  KEYS.N  KG.N^C11  \
Date                                                                        
1995-01-03      8.5625   6.5625    NaN     NaN  12.6250     NaN       NaN   
1995-01-04      8.8750   6.5000    NaN     NaN  12.6250     NaN       NaN   
1995-01-05      9.0000   6.6875    NaN     NaN  12.7500     NaN       NaN   
1995-01-06      9.0000   6.6875    NaN     NaN  12.8125     NaN       NaN   
1995-01-09      8.9375   6.8750    NaN     NaN  12.8750     NaN       NaN   
...                ...      ...    ...     ...      ...     ...       ...   
2024-12-24         NaN  66.4000  34.99   32.45  17.3700  163.86       NaN   
2024-12-26         NaN  66.2700  35.43   32.31  17.3900  164.63       NaN   
2024-12-27         NaN  65.6900  34.83   32.27  17.2000  162.91       NaN   
2024-12-30         NaN  65.5800  34.53   31.96  17.1100  161.17       NaN   
2024-12-31         NaN  65.7200  34.60   32.12  17.1400  160.63       NaN   

            KHC.OQ  
Date                
1995-01-03     NaN  
1995-01-04     NaN  
1995-01-05     NaN  
1995-01-06     NaN  
1995-01-09     NaN  
...            ...  
2024-12-24   30.64  
2024-12-26   30.55  
2024-12-27   30.68  
2024-12-30   30.36  
2024-12-31   30.71  

[7552 rows x 632 columns]

In [5]:
movers = pd.read_excel(mover_file, header=None)
movers.columns = ['Status', 'Issuer', 'Code', 'Date']
display(movers.head())

Status            Issuer      Code       Date
0  Leaver  Eastman Chemical    EMN.N  2025-11-04
1  Joiner  Qnity Electronic      Q.N  2025-11-03
2  Leaver            Carmax    KMX.N  2025-10-31
3  Joiner  Solstice Advance  SOLS.OQ  2025-10-30
4  Joiner          Applovin   APP.OQ  2025-09-22

In [6]:
movers['Date'] = pd.to_datetime(movers['Date'])
movers.sort_values(by='Date', inplace=True)
display(movers.head(5))

Status            Issuer                 Code       Date
2120  Joiner          eSystems  ESY.N^E95 (expired) 1994-12-30
1789  Joiner            Ecolab               ECL.N  1994-12-30
1790  Joiner       Edison Intl               EIX.N  1994-12-30
1791  Joiner       El Paso CGP  CGP.N^A01 (expired) 1994-12-30
1792  Joiner  Emerson Electric               EMR.N  1994-12-30

In [7]:
unique_codes = movers['Code'].unique()
unique_codes = [t.replace("(expired)", "").strip() for t in unique_codes]


In [8]:
december_1994_data = movers[movers['Date'].dt.to_period('M') == '1994-12']
current_sp500_composition = set(december_1994_data['Code'])
print(f"Size of the initial S&P 500 composition set: {len(current_sp500_composition)}")

Size of the initial S&P 500 composition set: 483


In [9]:
sp500_composition_over_time = {}

for index, row in movers.iterrows():
    date = row['Date']
    status = row['Status']
    code = row['Code']
    code = code.replace("(expired)", "").strip()

    if status == 'Joiner':
        current_sp500_composition.add(code)
    elif status == 'Leaver':
        try:
            current_sp500_composition.remove(code)
        except KeyError:
            # Handle cases where a leaver is not found in the current composition
            pass

    # Optionally store the composition at this date
    sp500_composition_over_time[date] = set(current_sp500_composition)

print(f"Number of dates with composition changes recorded: {len(sp500_composition_over_time)}")

Number of dates with composition changes recorded: 742


In [10]:
dates = sorted(sp500_composition_over_time.keys())
unique_months = sorted(list(set(date.to_period('M') for date in dates)))

monthly_sp500_composition = {}

for month in unique_months:
    latest_date_in_month = None
    for date in dates:
        if date.to_period('M') == month:
            latest_date_in_month = date

    if latest_date_in_month is not None:
        monthly_sp500_composition[month] = sp500_composition_over_time[latest_date_in_month]

print(f"Number of months with recorded composition: {len(monthly_sp500_composition)}")

Number of months with recorded composition: 309


In [11]:
all_codes = sorted(list(unique_codes))
# Generate all months from the first date in the data to the last date
start_month = movers['Date'].min().to_period('M')
end_month = movers['Date'].max().to_period('M')
all_months = pd.period_range(start=start_month, end=end_month, freq='M').tolist()


data = []
current_composition = set(december_1994_data['Code']) # Initialize with the December 1994 composition

# Recreate the sp500_composition_over_time dictionary with daily compositions
sp500_composition_over_time = {}
for index, row in movers.iterrows():
    date = row['Date']
    status = row['Status']
    code = row['Code']
    code = code.replace("(expired)", "").strip()

    if status == 'Joiner':
        current_composition.add(code)
    elif status == 'Leaver':
        try:
            current_composition.remove(code)
        except KeyError:
            pass # Handle cases where a leaver is not found in the current composition

    sp500_composition_over_time[date] = set(current_composition)


# Iterate through all months and find the latest composition for each
monthly_sp500_composition = {}
dates_with_composition = sorted(sp500_composition_over_time.keys())

for month in all_months:
    latest_date_in_month = None
    # Find the latest date in sp500_composition_over_time that is in the current month
    for date in reversed(dates_with_composition): # Iterate in reverse to find the latest date quickly
        if date.to_period('M') == month:
            latest_date_in_month = date
            break

    # If a date with composition is found in the month, use its composition
    if latest_date_in_month is not None:
        monthly_sp500_composition[month] = sp500_composition_over_time[latest_date_in_month]
    else:
        # If no date with composition change in the current month,
        # use the composition from the previous month's last recorded change.
        # This assumes the composition remains unchanged unless a joiner/leaver occurs.
        previous_month_composition = None
        for date in reversed(dates_with_composition):
             if date < month.start_time:
                previous_month_composition = sp500_composition_over_time[date]
                break
        if previous_month_composition is not None:
             monthly_sp500_composition[month] = previous_month_composition
        else:
            # This case should ideally not happen if the data starts in Dec 1994
            # but as a fallback, use the initial composition
             monthly_sp500_composition[month] = set(december_1994_data['Code'])


data = []
for month in all_months:
    month_data = {'Month': month.to_timestamp()}
    current_month_composition = monthly_sp500_composition.get(month, set())
    for code in all_codes:
        month_data[code] = 1 if code in current_month_composition else 0
    data.append(month_data)


sp500_composition_df = pd.DataFrame(data)
sp500_composition_df.set_index('Month', inplace=True)
display(sp500_composition_df.head())

A.N  AA.N  AABA.OQ^J19  AAL.N^B97  AAL.OQ  AAP.N  AAPL.OQ  ABBV.N  \
Month                                                                           
1994-12-01    0     0            0          1       0      0        1       0   
1995-01-01    0     0            0          1       0      0        1       0   
1995-02-01    0     0            0          1       0      0        1       0   
1995-03-01    0     0            0          1       0      0        1       0   
1995-04-01    0     0            0          1       0      0        1       0   

            ABI.N^K08  ABK.N^K10  ...  YNR.N^J00  YUM.N  YUMC.N  ZBH.N  \
Month                             ...                                    
1994-12-01          0          0  ...          0      0       0      0   
1995-01-01          0          0  ...          0      0       0      0   
1995-02-01          0          0  ...          0      0       0      0   
1995-03-01          0          0  ...          0      0       0      0   
1995-04-01          0          0  ...          0      0       0      0   

            ZBRA.OQ  ZE.N^E98  ZIMV.OQ^J25  ZION.OQ  ZRN.N^F98  ZTS.N  
Month                                                                  
1994-12-01        0         1            0        0          1      0  
1995-01-01        0         1            0        0          1      0  
1995-02-01        0         1            0        0          1      0  
1995-03-01        0         1            0        0          1      0  
1995-04-01        0         1            0        0          1      0  

[5 rows x 1268 columns]

Check which stocks are part of the index at the end of the training period. These stocks will be considered for training and testing. We get all available return data for these stocks in the 4 years of the study period (also if they later joined S&P500, we get their price data for the whole training period). Also, if companies leave S&P500, we still use them in the test set as long as we have price data available.

In [12]:
i = 0
study_periods_folder = notebook_dir / 'data' / 'study_periods'
for year in range(1997, 2024):
    i += 1
    study_period_series = []  # Collect series here
    closest_date = sp500_composition_df.index.get_indexer([pd.Timestamp(f'{year}-12-01')], method='nearest')[0]
    study_period_stocks = sp500_composition_df.iloc[closest_date]
    # only keep 1s
    study_period_stocks = study_period_stocks[study_period_stocks == 1] 
    study_period_stocks = study_period_stocks.index.tolist()
    # get returns for each constituent for the last 3 years and current year
    for stock in study_period_stocks:
        try:
            stock_returns = returns[stock][f'{year-2}-01-01':f'{year+1}-12-31']
            study_period_series.append(stock_returns.rename(stock))  # Rename to stock ticker
        except KeyError:
            continue
    # Concat all series at once
    study_period = pd.concat(study_period_series, axis=1)
    study_period.to_csv(study_periods_folder / f'test_{year+1}.csv')

print(f"Generated {i} study periods.")

Generated 27 study periods.


In [13]:
data = pd.read_csv(study_periods_folder / 'test_2005.csv')
data['Date'] = pd.to_datetime(data['Date'])
data.set_index('Date', inplace=True)

In [14]:
missing_value_stocks = data.columns[data.isna().sum(axis=0) > 0]
# print number of na values for stocks with missing values
for stock in missing_value_stocks:
    print(f"{stock}: {data[stock].isna().sum()} missing values")
    if data[stock].isna().sum() == len(data):
        # delete column
        data.drop(columns=stock, inplace=True)
        print('Deleted column:', stock)

AMCR.N: 1008 missing values
Deleted column: AMCR.N
CIT.N^K09: 125 missing values
CPN.N^L05: 18 missing values
DAL.N^J05: 55 missing values
DPH.N^J05: 57 missing values
FSLb.N^L06: 729 missing values
G.N^J05: 63 missing values
GENZ.OQ^D11: 1 missing values
GILD.OQ: 1 missing values
GLK.N^G05: 126 missing values
GP.N^L05: 5 missing values
HSP.N^I15: 578 missing values


NOTE: If missing values are >250, they cannot occur at the very end. For stocks with missing values, either only 1000 - #NA sequences are created (if missing at the beginning) or only #days_of_year - #NA predictions will be generated (if missing at the end)

In [15]:
# create returns out of the prices
returns_data = data.pct_change()
returns_data = returns_data[1:]
returns_data

C:\Users\jdobe\AppData\Local\Temp\ipykernel_16208\4225891145.py:2: FutureWarning: The default fill_method='pad' in DataFrame.pct_change is deprecated and will be removed in a future version. Either fill in any non-leading NA values prior to calling pct_change or specify 'fill_method=None' to not fill NA values.
  returns_data = data.pct_change()


A.N  AABA.OQ^J19   AAPL.OQ  ABI.N^K08  ABK.N^K10  ABS.N^F06  \
Date                                                                           
2002-01-03  0.063248     0.026838  0.012017  -0.051053  -0.007031  -0.002534   
2002-01-04  0.054019    -0.012023  0.004664   0.046866   0.007427   0.011115   
2002-01-07 -0.003966     0.043915 -0.033347  -0.009272  -0.007543  -0.016960   
2002-01-08  0.003063    -0.010137 -0.012662  -0.050802  -0.012956  -0.021406   
2002-01-09 -0.023817     0.036866 -0.042459   0.021127   0.010326  -0.007835   
...              ...          ...       ...        ...        ...        ...   
2005-12-23 -0.004847    -0.005143 -0.009052  -0.011940   0.007436  -0.117698   
2005-12-27 -0.026361    -0.016741  0.011997  -0.003021  -0.005981   0.013632   
2005-12-28 -0.002354     0.007261 -0.008892   0.013258  -0.006657   0.013449   
2005-12-29 -0.012684    -0.016654 -0.028816   0.001495  -0.000773   0.013744   
2005-12-30 -0.005378    -0.009606  0.006158  -0.008585  -0.006062  -0.001870   

               ABT.N  ACS.N^B10  ACV.N^K06   ADBE.OQ  ...  JNS.N^E17  \
Date                                                  ...              
2002-01-03  0.000895  -0.021382   0.003868  0.038945  ...   0.021466   
2002-01-04 -0.001789   0.017188  -0.005213  0.085248  ...   0.047464   
2002-01-07 -0.006990  -0.007160  -0.010025  0.007242  ...  -0.007610   
2002-01-08 -0.006498   0.017308  -0.007135  0.013274  ...  -0.015336   
2002-01-09 -0.005451   0.020605   0.010431  0.027020  ...   0.014159   
...              ...        ...        ...       ...  ...        ...   
2005-12-23  0.002946   0.050275   0.004354 -0.009079  ...   0.007483   
2005-12-27 -0.011258  -0.021639  -0.011706 -0.004581  ...  -0.008488   
2005-12-28 -0.001238  -0.003854   0.012064 -0.001895  ...   0.003210   
2005-12-29 -0.013631  -0.002523   0.000650  0.002983  ...   0.001067   
2005-12-30 -0.009296  -0.002024  -0.009097 -0.000541  ...  -0.007459   

            JNY.N^D14  JP.N^D06     JPM.N  JWN.N^E25       K.N  KATE.N^G17  \
Date                                                                         
2002-01-03   0.003322  0.003272  0.026685   0.035158  0.000667    0.023023   
2002-01-04   0.035220  0.002392  0.045016   0.021349 -0.021340    0.027789   
2002-01-07  -0.016865 -0.010195 -0.002564  -0.011401  0.003407    0.009139   
2002-01-08   0.013014  0.006355 -0.007712   0.074964  0.012224   -0.000189   
2002-01-09  -0.018394  0.008711  0.002850  -0.033974 -0.002348   -0.004718   
...               ...       ...       ...        ...       ...         ...   
2005-12-23   0.004557  0.004545  0.006492   0.003209 -0.005185    0.015710   
2005-12-27  -0.004213 -0.003481 -0.005210  -0.018662 -0.006572   -0.001969   
2005-12-28   0.005857  0.003493 -0.004738   0.022277 -0.001597    0.014088   
2005-12-29  -0.006147 -0.005743 -0.001253  -0.004252 -0.006626   -0.003890   
2005-12-30   0.000000 -0.003501 -0.004265  -0.001868 -0.005980   -0.000837   

               KBH.N     KEY.N  KG.N^C11  
Date                                      
2002-01-03 -0.011367  0.004082 -0.025788  
2002-01-04 -0.003321  0.007724 -0.019608  
2002-01-07 -0.001282 -0.001614  0.004500  
2002-01-08 -0.013090 -0.006869  0.009209  
2002-01-09  0.014564 -0.002441 -0.015783  
...              ...       ...       ...  
2005-12-23 -0.013898  0.000592 -0.001763  
2005-12-27 -0.003356 -0.010947  0.019423  
2005-12-28 -0.017374 -0.004786 -0.002309  
2005-12-29  0.002193 -0.003306 -0.017361  
2005-12-30 -0.006291 -0.006936 -0.003534  

[1007 rows x 258 columns]

In [16]:
# Only use data from first 3 years for normalization
start_date = returns_data.index.min().year
end_date = start_date + 2
start_date = f'{start_date}-01-01'
end_date = f'{end_date}-12-31'

returns_3years = returns_data[f'{start_date}':f'{end_date}']
mean_return = returns_3years.mean().mean()
mean_volatility = returns_3years.std().mean()
returns_data = (returns_data - mean_return) / mean_volatility

In [17]:
# stack returns in one large vector
stacked_returns = returns_data.T.stack().reset_index()
stacked_returns.columns = ['stock', 'date', 'return']
stacked_returns

stock       date    return
0            A.N 2002-01-03  2.739748
1            A.N 2002-01-04  2.336460
2            A.N 2002-01-07 -0.197491
3            A.N 2002-01-08  0.109663
4            A.N 2002-01-09 -1.064977
...          ...        ...       ...
258369  KG.N^C11 2005-12-23 -0.101209
258370  KG.N^C11 2005-12-27  0.824611
258371  KG.N^C11 2005-12-28 -0.125106
258372  KG.N^C11 2005-12-29 -0.782862
258373  KG.N^C11 2005-12-30 -0.178599

[258374 rows x 3 columns]

In [18]:
window = 240  # number of past days in each sequence

# Group by stock and create rolling windows
sequences = []

for stock, group in stacked_returns.groupby('stock'):
    vals = group['return'].values
    
    # create rolling sequences of length 240
    for i in range(window, len(vals)):
        seq = vals[i-window:i]  # past 240 returns
        sequences.append({
            'stock': stock,
            'date': group['date'].iloc[i],  # the "current" date (t)
            'return': vals[i],  # the return at time t
            'sequence': seq
        })

# Convert to DataFrame
sequences_df = pd.DataFrame(sequences)

## Targets
for each date, get the returns for all stocks and create two targets: below median and above median return


In [19]:
# get returns for all stocks for a specific date (day)
def get_returns_for_date(returns_df, target_date):
    """
    Get returns for all stocks for a specific date.

    Parameters:
    returns_df (pd.DataFrame): DataFrame with Date as index and stock tickers as columns.
    target_date (str or pd.Timestamp): The date for which to retrieve returns.

    Returns:
    pd.Series: Series with stock tickers as index and their returns on the target date.
    """
    target_date = pd.to_datetime(target_date)
    if target_date not in returns_df.index:
        raise ValueError(f"No data available for the date: {target_date}")
    else:
        return returns_df.loc[target_date]

In [20]:
from datetime import datetime
from dateutil.relativedelta import relativedelta

end = (datetime.strptime(end_date, '%Y-%m-%d') + relativedelta(years=1)).strftime('%Y-%m-%d')

valid_days = returns_data.loc[
    (returns_data.index >= start_date) & 
    (returns_data.index <= end)
].index

# Initialize target columns with 0 (integer type)
sequences_df['Class0'] = 0
sequences_df['Class1'] = 0

for date in valid_days:
    current = date.strftime('%Y-%m-%d')
    returns_on_date = get_returns_for_date(returns_data, current)
    # compute median return
    median_return = returns_on_date.median()
    # create two targets: below median and above median
    below_median = returns_on_date[returns_on_date < median_return].index
    above_median = returns_on_date[returns_on_date >= median_return].index
    # add these results to the sequences_df
    mask_current_date = sequences_df['date'] == current

    sequences_df.loc[
        mask_current_date & sequences_df['stock'].isin(below_median),
        'Class0'
    ] = 1

    sequences_df.loc[
        mask_current_date & sequences_df['stock'].isin(above_median),
        'Class1'
    ] = 1

In [21]:
sequences_df

stock       date    return  \
0            A.N 2002-12-16  1.042911   
1            A.N 2002-12-17 -0.569814   
2            A.N 2002-12-18 -1.405504   
3            A.N 2002-12-19 -0.257593   
4            A.N 2002-12-20  3.208991   
...          ...        ...       ...   
196449  KG.N^C11 2005-12-23 -0.101209   
196450  KG.N^C11 2005-12-27  0.824611   
196451  KG.N^C11 2005-12-28 -0.125106   
196452  KG.N^C11 2005-12-29 -0.782862   
196453  KG.N^C11 2005-12-30 -0.178599   

                                                 sequence  Class0  Class1  
0       [2.7397480610005727, 2.3364597617248255, -0.19...       0       1  
1       [2.3364597617248255, -0.1974907018107239, 0.10...       1       0  
2       [-0.1974907018107239, 0.1096632485976826, -1.0...       1       0  
3       [0.1096632485976826, -1.0649772347946689, -0.3...       0       1  
4       [-1.0649772347946689, -0.36590736053518036, -0...       0       1  
...                                                   ...     ...     ...  
196449  [-3.0128376956664535, -0.7675122093278653, 0.3...       1       0  
196450  [-0.7675122093278653, 0.3340141846843717, -0.3...       0       1  
196451  [0.3340141846843717, -0.3794666513231489, -0.2...       1       0  
196452  [-0.3794666513231489, -0.22318037295959992, -0...       1       0  
196453  [-0.22318037295959992, -0.10414567102925175, -...       0       1  

[196454 rows x 6 columns]

In [22]:
sequences_df['Class0'].sum(), sequences_df['Class1'].sum()

(np.int64(98111), np.int64(98343))

In [23]:
sequences_df.isna().sum()

stock       0
date        0
return      0
sequence    0
Class0      0
Class1      0
dtype: int64

## LSTM Network

- the use RMSprop instead of adam
- recurrent dropout = 0.1
- 80% of training samples as training set and 20% as validation set
- early stopping, with patience = 10, while max epochs = 1000
- 1 input with 240 lags
- 1 LSTM layer with h = 25
- dense output layer with 2 neurons and softmax activation function

Forecasting goal: probability for each stock to under/outperform the cross-sectional median at t+1 with information up to point t. Stocks are ranked in descending order for this probability for each of the 2 targets.

In [33]:
import os
os.environ['KERAS_BACKEND'] = 'torch'

import numpy as np
import torch
from sklearn.model_selection import train_test_split
from keras.models import Sequential
from keras.layers import LSTM, Dense
from keras.callbacks import EarlyStopping

In [34]:
# Prepare the data
X = np.array(sequences_df['sequence'].tolist())
X = X.reshape(X.shape[0], 240, 1)
y = sequences_df[['Class0', 'Class1']].values

# Split into training and validation sets (80% train, 20% val)
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)

In [ ]:
# Build the LSTM model
model = Sequential()
model.add(LSTM(25, input_shape=(240, 1), recurrent_dropout=0.1))
model.add(Dense(2, activation='softmax'))

# Compile the model
model.compile(optimizer='rmsprop', loss='categorical_crossentropy', metrics=['accuracy'])

# Display the model summary
model.summary()

# Move model to GPU if available
device = 'cuda' if torch.cuda.is_available() else 'cpu'
print(f'Using device: {device}')
model.to(device)

Model: "sequential_3"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ lstm_3 (LSTM)                   │ (None, 25)             │         2,700 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_3 (Dense)                 │ (None, 2)              │            52 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 2,752 (10.75 KB)

 Trainable params: 2,752 (10.75 KB)

 Non-trainable params: 0 (0.00 B)

Using device: cpu


<Sequential name=sequential_3, built=True>

In [ ]:
# Set up early stopping
early_stop = EarlyStopping(monitor='val_loss', patience=10)

# Train the model
history = model.fit(X_train, y_train, epochs=100, validation_data=(X_val, y_val), callbacks=[early_stop])

Epoch 1/1000
4912/4912 ━━━━━━━━━━━━━━━━━━━━ 1319s 268ms/step - accuracy: 0.5008 - loss: 0.6934 - val_accuracy: 0.5032 - val_loss: 0.6932
Epoch 2/1000
4912/4912 ━━━━━━━━━━━━━━━━━━━━ 1319s 268ms/step - accuracy: 0.5008 - loss: 0.6934 - val_accuracy: 0.5032 - val_loss: 0.6932
Epoch 2/1000
4912/4912 ━━━━━━━━━━━━━━━━━━━━ 1198s 244ms/step - accuracy: 0.5048 - loss: 0.6932 - val_accuracy: 0.5057 - val_loss: 0.6930
Epoch 3/1000
4912/4912 ━━━━━━━━━━━━━━━━━━━━ 1198s 244ms/step - accuracy: 0.5048 - loss: 0.6932 - val_accuracy: 0.5057 - val_loss: 0.6930
Epoch 3/1000
4912/4912 ━━━━━━━━━━━━━━━━━━━━ 1122s 228ms/step - accuracy: 0.5023 - loss: 0.6932 - val_accuracy: 0.5031 - val_loss: 0.6931
Epoch 4/1000
4912/4912 ━━━━━━━━━━━━━━━━━━━━ 1122s 228ms/step - accuracy: 0.5023 - loss: 0.6932 - val_accuracy: 0.5031 - val_loss: 0.6931
Epoch 4/1000
3280/4912 ━━━━━━━━━━━━━━━━━━━━ 5:51 216ms/step - accuracy: 0.5041 - loss: 0.6931

KeyboardInterrupt: 